taxcalc/utils.py

```diff
@@ -231,33 +235,66 @@ def means_and_comparisons(df, col_name, gp, weighted_total):
 
     return diffs
 
+def weighted(df, X):
+    agg = df
+    for colname in X:
+        if colname != 's006':
+            agg[colname] = df[colname]*df['s006']
+    return agg
+
 
 def results(c):
     outputs = [getattr(c, col) for col in STATS_COLUMNS]
     return DataFrame(data=np.column_stack(outputs), columns=STATS_COLUMNS)
 
 
-def create_distribution_table(calc, groupby):
+def weighted_avg_allcols(df, cols):    
+    diff = DataFrame(df.groupby('bins').apply(weighted_mean, "c00100"), columns = ['c00100'])
+    
+    for col in cols:
+        if col == "s006" or col =='num_returns_StandardDed' or col == 'num_returns_ItemDed' or col == 'num_returns_AMT':
+            diff[col] = df.groupby('bins')[col].sum()
+        elif col != "c00100":
+            diff[col] = df.groupby('bins').apply(weighted_mean, col)
+            
+    return diff
+
+def create_distribution_table(calc, groupby, result_type):
```

Comment:
This function is definitely an improvement over the previous version, but I think there might be a nicer way to write the math. I'll have a go at it.